packages/did/src/Did.signature.ts

```diff
@@ -59,23 +62,46 @@ function verifyDidSignatureDataStructure(
  * @param input.message The message that was signed.
  * @param input.signature Signature bytes.
  * @param input.keyUri DID URI of the key used for signing.
+ * @param input.expectedSigner If given, verification fails if the controller of the signing key is not the expectedSigner.
+ * @param input.allowUpgraded If `expectedSigner` is a light DID, setting this flag to `true` will accept signatures by the corresponding full DID.
  * @param input.expectedVerificationMethod Which relationship to the signer DID the key must have.
  * @param input.didResolveKey Allows specifying a custom DID key resolve. Defaults to the built-in [[resolveKey]].
  */
 export async function verifyDidSignature({
   message,
   signature,
   keyUri,
+  expectedSigner,
+  allowUpgraded = false,
   expectedVerificationMethod,
   didResolveKey = resolveKey,
 }: DidSignatureVerificationInput): Promise<void> {
   // Verification fails if the signature key URI is not valid
-  const { fragment } = parse(keyUri)
-  if (!fragment)
+  const signer = parse(keyUri)
+  if (!signer.fragment)
     throw new SDKErrors.SignatureMalformedError(
       `Signature key URI "${keyUri}" invalid`
     )
 
+  // checks if key uri points to the right did; alternatively we could check the key's controller
+  if (expectedSigner && expectedSigner !== signer.did) {
+    // check for allowable exceptions
+    const expected = parse(expectedSigner)
+    // NECESSARY CONDITION: subjects and versions match
+    const subjectVersionMatch =
+      expected.address === signer.address && expected.version === signer.version
+    // EITHER: signer is a full did and we allow signatures by corresponding full did
+    const allowedUpgrade = allowUpgraded && signer.type === 'full'
+    // OR: both are light dids and their auth key type matches
+    const keyTypeMatch =
+      signer.type === 'light' &&
+      expected.type === 'light' &&
```

Comment:
I'm not sure these two checks are necessary - a key type match could happen if both dids are full dids and do not have a key type indicator, but in that case identical addresses could only mean they are the same did?

Reply:
Indeed, it looks like checking for the light type can be skipped